puppet/modules/jenkins_job_builder/files/theforeman.org/pipelines/release/foreman-deb.groovy

```diff
@@ -10,100 +10,12 @@ pipeline {
-                            deleteDir()
-                        }
-                    }
-                }
-
-                stage('ubuntu1604') {
-                    agent { label 'el' }
-                    environment {
-                        cico_job_name = "foreman-nightly-ubuntu1604-test"
-                    }
-
-                    steps {
-                        git_clone_foreman_infra()
-
-                        withCredentials([string(credentialsId: 'centos-jenkins', variable: 'PASSWORD')]) {
-                            runPlaybook(
-                                playbook: 'ci/centos.org/ansible/jenkins_job.yml',
-                                extraVars: [
-                                    "jenkins_job_name": "${env.cico_job_name}",
-                                    "jenkins_username": "foreman",
-                                    "jenkins_job_link_file": "${env.WORKSPACE}/jobs/${env.cico_job_name}"
-                                ],
-                                sensitiveExtraVars: ["jenkins_password": "${env.PASSWORD}"]
-                            )
-                        }
-                    }
-                    post {
-                        always {
-                            script {
-                                set_job_build_description("${env.cico_job_name}")
-                            }
-                            deleteDir()
-                        }
-                    }
-                }
-
-                stage('ubuntu1804') {
-                    agent { label 'el' }
-                    environment {
-                        cico_job_name = "foreman-nightly-ubuntu1804-test"
-                    }
-
-                    steps {
-                        git_clone_foreman_infra()
-
-                        withCredentials([string(credentialsId: 'centos-jenkins', variable: 'PASSWORD')]) {
-                            runPlaybook(
-                                playbook: 'ci/centos.org/ansible/jenkins_job.yml',
-                                extraVars: [
-                                    "jenkins_job_name": "${env.cico_job_name}",
-                                    "jenkins_username": "foreman",
-                                    "jenkins_job_link_file": "${env.WORKSPACE}/jobs/${env.cico_job_name}"
-                                ],
-                                sensitiveExtraVars: ["jenkins_password": "${env.PASSWORD}"]
-                            )
-                        }
-                    }
-                    post {
-                        always {
-                            script {
-                                set_job_build_description("${env.cico_job_name}")
-                            }
-                            deleteDir()
-                        }
-                    }
-                }
+            steps {
+                parallel(
+                    'debian9': { runCicoJob("foreman-nightly-debian9-test", 0) },
```

Comment:
This call signature is a little ugly and I think it could be made cleaner in one of two ways:

 1) make a `runCicoJobsInParallel` function that handles the numbering and creation based on a hash
 2) Make this a for loop (#1 includes #2)

Reply:
I experimented with this. Thought in this PR it was a bit overkill and premature, but https://github.com/theforeman/foreman-infra/pull/992/files#diff-ef3be3111b17e7d2f2296bc04cd28ec8R45

Reply:
That logic wrapped in a function is exactly what I am thinking. Take a look as well at https://github.com/SatelliteQE/robottelo-ci/blob/master/workflows/lib/runPlaybook.groovy#L51

Reply:
I don't see a way to clean it up so it's more readable. Would you mind if we merge it as it is (after some testing) and iterate on it in a later stage?

Reply:
Sure. Here's what I envision just in case:

```
def runCicoJobInParallel(jobs) {

    for (int i = 0; i < jobs.size(); i++) {
        def index = i // fresh variable per iteration; i will be mutated
        branches[jobs[i]['name'] = {
            runCicoJob(jobs[i]['job'], i)
        }
    }

    parallel branches
}
```

```
steps {
    script {
        runCicoJobInParallel([
            ['name': 'ubuntu1804', 'job': 'foreman-nightly-ubuntu1804-test'],
            ['name': 'ubuntu1604', 'job': 'foreman-nightly-ubuntu1604-test']
        ])
    }
}
```

Reply:
The problem is that if you start to add the additional vars to also set the OS it becomes ugly. Maybe that could be solved by defining multiple cico jobs via templates.

Another thought: can we pass in a string template? So the call could be:

```groovy
steps {
    script {
        runCicoJobInParallel('foreman-nightly-{}-test', ['ubuntu1804', 'ubuntu1604'])
    }
}
```

https://stackoverflow.com/questions/45830294/can-groovy-string-interpolation-be-nested suggests this may be possible but not sure if we're overengineering things then.

I think I'll implement your solution since it at least hides the counter from the actual job definition.